perses/analysis/free_energies.py

```diff
@@ -0,0 +1,242 @@
+        Path to json file containing run metadata.
+    work_file_path : str
+        Path to bz2-compressed pickle file containing work values from simulation
+    complex_project : str
+        Project identifier (of the form "PROJXXXXX") of the FAH project for complex simulation
+    solvent_project : str
+        Project identifier (of the form "PROJXXXXX") of the FAH project for solvent simulation
+    temperature_kelvin : float
+    n_bootstrap : int, optional
+        Number of bootstrap steps used for BAR free energy estimation
+    show_plots : bool, optional
+        If true, block to display plots interactively (using `plt.show()`).
+        If false, save plots as images in current directory (for batch usage).
+    plot_file_format : str, optional
+        Image file format for saving plots. Accepts any file extension supported by `plt.savefig()`
+    cache_dir : str, optional
+        If given, local directory for caching BAR calculations. Results are cached by run, phase, and generation.
+        If None, no caching is performed.
+    """
+
+    with bz2.BZ2File(work_file_path, "r") as infile:
+        work = pickle.load(infile)
+
+    with open(details_file_path, "r") as f:
+        details = json.load(f)
+
+    work.replace([np.inf, -np.inf], np.nan)
+    work = work.dropna()
+
+    temperature = temperature_kelvin * unit.kelvin
+    kT = kB * temperature
+
+    projects = {"complex": complex_project, "solvent": solvent_project}
+
+    def _produce_plot(name):
+        if show_plots:
+            plt.show()
+        else:
+            plt.savefig('.'.join([name, plot_file_format]))
+
+    def _bootstrap_BAR(run, phase, gen_id):
+        f_works, r_works = _get_works(work, RUN, projects[phase], GEN=f"GEN{gen_id}")
+        f_works = _strip_outliers(f_works)
+        r_works = _strip_outliers(r_works)
+        fes = []
+        errs = []
+
+        if len(f_works) > 10 and len(r_works) > 10:
+            for _ in range(n_bootstrap):
+                f = random.choices(f_works, k=len(f_works))
+                r = random.choices(r_works, k=len(r_works))
+                fe, err = BAR(np.asarray(f), np.asarray(r))
+                fes.append(fe)
+                errs.append(err)
+
+        return fes, errs, f_works, r_works
+
+    if cache_dir is not None:
+        # cache results in local directory
+        memory = joblib.Memory(backend="local", location=cache_dir, verbose=0)
+        bootstrap_BAR = memory.cache(_bootstrap_BAR)
+    else:
+        bootstrap_BAR = _bootstrap_BAR
+
+    for d in tqdm.tqdm(details.values()):
+        RUN = d["directory"]
+        if show_plots:
+            fig, axes = plt.subplots(ncols=2, nrows=1, figsize=(10, 5))
+        for i, phase in enumerate(projects.keys()):
+            if show_plots:
+                axes[i].set_title(phase)
+
+            all_forward = []
+            all_reverse = []
+            # There will be 6 gens for this project I think
+            for gen_id in range(0, 7):
```

Comment:
I wrote this better in the `ANALYSIS-prospective.ipynb`
so 7 isn't hard coded, but pulled from the dataframe itself

Reply:
I adapted this in 01a88c7 (just changed it to get the maximum gen by run, since I wasn't sure if all runs are guaranteed to have the same number of gens).